[Hub Generated] Review request for Microsoft.ApiManagement to add version preview/2019-12-01-preview

solankisamir opened the pull request:
If you are a MSFT employee you can view your [work branch via this link](https://portal.azure-devex-tools.com/branch/solankisamir/azure-rest-api-specs/solanki%2Fapim-2019-12-01-preview...master/).

### Contribution checklist:
- [x] I have reviewed the [documentation](https://github.com/Azure/azure-rest-api-specs#basics) for the workflow.
- [ ] [Validation tools](https://github.com/Azure/azure-rest-api-specs/blob/master/documentation/swagger-checklist.md#validation-tools-for-swagger-checklist) were run on swagger spec(s) and have all been fixed in this PR.

solankisamir committed:
Adds base for updating Microsoft.ApiManagement from version stable/2019-01-01 to version 2019-12-01-preview

solankisamir committed:
Updates readme

solankisamir committed:
Updates API version in new specs and examples

vfedonkin committed:
ApiManagement - renaming /properties entity to /namedValues (#7683)

* ApiManagement - renaming /properties entity to /namedValues

* ApiManagement - renaming parameter name: propId to namedValueId

vfedonkin committed:
Apim 2019 12 01 preview (#7705)

* ApiManagement - async operatins for NamedValues and ApiSchemas

* ApiManagement - client secrets in IdentityProvider

* ApiManagement - json format fixes

* ApiManagement - examples fixes

* ApiManagement - fixed identtyProviderContract

* ApiManagement - secrets in subscriptions

* ApiManagement - secrets in OpenIdConnectProviders and AuthServers

* ApiManagement - secrets in TenantAccess

* ApiManagement - fix example

* ApiManagement - fixes

vfedonkin committed:
Apim 2019 12 01 preview (#7725)

* ApiManagement - tagDescription id

* ApiManagement - policyDescriptions

* ApiManagement - policyDescriptions small fix

* ApiManagement - policyDescriptions sample

* ApiManagement - revisions id and api schema doc

* ApiManagement - tag descriptions list example

* ApiManagement - ciphers info

vfedonkin committed:
Rpchanges in 2019-12-01-preview (#8039)

* update error response for apim deployment

* disable gateway

* user assigned identities

* api version contraint

* update examples

* fix model issues in examples

* update example

* replace sasolankrp with apimService1

Co-authored-by: Samir Solanki <samirsolanki@outlook.com>

solankisamir committed:
prettier check update

AutorestCI commented:
/azp run automation - sdk

AutorestCI commented:
/azp run automation - sdk

solankisamir commented:
This is PR related to https://github.com/Azure/azure-rest-api-specs/pull/8046. We couldn't the apim-2019-12-01-preview branch as it was read-only. Hence created a clone here and working on edits here.

solankisamir committed:
second attempt at prettier

AutorestCI commented:
/azp run automation - sdk

solankisamir commented:
@NullMDR I am trying to fix the prettier issues using https://github.com/Azure/azure-rest-api-specs/blob/master/documentation/ci-fix.md, but I am getting this error 

> D:\github\azure-rest-api-specs>npm install prettier-fix
npm ERR! code E404
npm ERR! 404 Not Found - GET https://registry.npmjs.org/prettier-fix - Not found
npm ERR! 404
npm ERR! 404  'prettier-fix@latest' is not in the npm registry.
npm ERR! 404 You should bug the author to publish it (or use the name yourself!)
npm ERR! 404
npm ERR! 404 Note that you can also install from a
npm ERR! 404 tarball, folder, http url, or git url.

> npm ERR! A complete log of this run can be found in:
npm ERR!     C:\Users\sasolank\AppData\Roaming\npm-cache\_logs\2020-01-05T19_20_03_381Z-debug.log

NullMDR commented:
@solankisamir You need 

npm install; npm run prettier -- --write "specification/apimanagement/**/*.json"

solankisamir commented:
@NullMDR ran the command and hit the issue below. I dont have the `azure-rest-api-specs-test` repos locally. Do I need to get that?
[detailed_log](https://gist.github.com/solankisamir/5266d504d167221744e3f8250ce4a1f5)

> PS D:\github\azure-rest-api-specs> npm install; npm run prettier -- --write "specification/apimanagement/**/*.json"

> azure-rest-api-specs-tests@0.1.0 postinstall D:\github\azure-rest-api-specs
> scripts/switch-to-preproduction.sh
npm ERR! errno 1
npm ERR! azure-rest-api-specs-tests@0.1.0 postinstall: `scripts/switch-to-preproduction.sh`
npm ERR! Exit status 1
npm ERR!
npm ERR! Failed at the azure-rest-api-specs-tests@0.1.0 postinstall script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.

> npm ERR! A complete log of this run can be found in:
npm ERR!     C:\Users\sasolank\AppData\Roaming\npm-cache\_logs\2020-01-06T16_40_16_473Z-debug.log
npm ERR! missing script: prettier

> npm ERR! A complete log of this run can be found in:
npm ERR!     C:\Users\sasolank\AppData\Roaming\npm-cache\_logs\2020-01-06T16_40_18_669Z-debug.log

ryansbenson reviewed (commented):
One minor change

vfedonkin commented on specification/apimanagement/resource-manager/Microsoft.ApiManagement/preview/2019-12-01-preview/definitions.json:
```diff
@@ -0,0 +1,5157 @@
+    },
+    "ApiRevisionContract": {
+      "properties": {
+        "id": {
```
It's on the top level, not inside "properties".

solankisamir replied:
@ryansbenson please see @vfedonkin comment. This is a contract associated with a proxy resource and we are just adding `id` to fix the ARM warnings generated as part of live traffic.

vfedonkin replied:
@ryansbenson, @solankisamir , I think we will need to modify all entities to have our fields inside "properties", but it should be done in our next version, since we already behind the schedule for the version we trying to release.

solankisamir replied:
@ryansbenson a gentle ping on this!

solankisamir replied:
@ryansbenson we fixed this in the spec to unblock it.

NullMDR commented:
@solankisamir Please merge from latest master to pick up latest scripts and try again

solankisamir committed:
ran prettier script

AutorestCI commented:
/azp run automation - sdk

mmyyrroonn reviewed (approved):
LGTM. wait for ARM's signoff

azuresdkci commented:
Can one of the admins verify this patch?

solankisamir committed:
httpcorrelation protocol update

solankisamir committed:
open api json document import and export

AutorestCI commented:
/azp run automation - sdk

solankisamir committed:
removed id definition

solankisamir committed:
fixed example for tooling.

AutorestCI commented:
/azp run automation - sdk

solankisamir committed:
removed space

solankisamir commented on specification/apimanagement/resource-manager/readme.md:
````diff
@@ -28,15 +28,55 @@ These are the global settings for the ApiManagement API.
 These settings apply only when `--tag=package-2019-01` is specified on the command line.
 
-```yaml $(tag) == 'package-2019-01'
+``` yaml $(tag) == 'package-2019-01'
````
fixed

AutorestCI commented:
/azp run automation - sdk

AutorestCI commented:
/azp run automation - sdk

MyronFanQiu commented:
/azp run automation - sdk

MyronFanQiu commented:
@ryansbenson sorry! I made a mistake to merge it. Could you please review this PR, if still need some change we can open another or revert this one. @solankisamir Sorry. Please hold on the following process since ARM didn't sign off yet. Thanks.

NullMDR commented:
@MyronFanQiu You can create a PR to revert this one

sanjaiganesh commented:
>           }


For all async operations, add x-ms-long-running-operation:true and

, add "x-ms-long-running-operation-options" extension with "final-state-via" property.
``` json
"x-ms-long-running-operation-options": {
  "final-state-via": "location"
}
```

or

``` json
"x-ms-long-running-operation-options": {
  "final-state-via": "azure-async-operation"
}
```

---
Refers to: specification/apimanagement/resource-manager/Microsoft.ApiManagement/preview/2019-12-01-preview/apimapis.json:2361 in 13fb704. [](commit_id = 13fb70433aee59b3ec37ef3f54a162530abca46a, deletion_comment = False)

sanjaiganesh commented on specification/apimanagement/resource-manager/Microsoft.ApiManagement/preview/2019-12-01-preview/apimnamedvalues.json:
```diff
@@ -0,0 +1,388 @@
+        "responses": {
+          "202": {
+            "description": "Request to update NamedValue was accepted.",
+            "headers": {
```
For all async operations, add x-ms-long-running-operation:true and

, add "x-ms-long-running-operation-options" extension with "final-state-via" property.
``` json
"x-ms-long-running-operation-options": {
  "final-state-via": "location"
}
```

or

``` json
"x-ms-long-running-operation-options": {
  "final-state-via": "azure-async-operation"
}
```

sanjaiganesh commented on specification/apimanagement/resource-manager/Microsoft.ApiManagement/preview/2019-12-01-preview/apimnamedvalues.json:
```diff
@@ -0,0 +1,388 @@
+  "paths": {
+    "/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.ApiManagement/service/{serviceName}/namedValues": {
+      "get": {
+        "tags": [
```
missing ListBySubscrition and ListByResourceGroup calls

vfedonkin replied:
We don't have such methods, "ListByService" only

sanjaiganesh commented on specification/apimanagement/resource-manager/Microsoft.ApiManagement/preview/2019-12-01-preview/apimnamedvalues.json:
```diff
@@ -0,0 +1,388 @@
+          }
+        ],
+        "responses": {
+          "202": {
```
add sync response 200 as well.. as service could respond synchronously if no changes.

vfedonkin replied:
We use 204 for "update was successful" since no any data returned

sanjaiganesh commented on specification/apimanagement/resource-manager/Microsoft.ApiManagement/preview/2019-12-01-preview/apimnamedvalues.json:
```diff
@@ -0,0 +1,388 @@
+              "Location": {
+                "description": "The URL where the status of the long running operation can be checked.",
+                "type": "string"
+              }
```
check other comments on LRO tracking extension annotation

vfedonkin replied:
@sanjaiganesh , should we remove this header description and use this instead?:

"x-ms-long-running-operation-options": {
  "final-state-via": "location"
}

sanjaiganesh commented:
looks fine.. added few comments. please address them in separate PR. added armsignoff.

solankisamir commented:
> looks fine.. added few comments. please address them in separate PR. added armsignoff.

@sanjaiganesh Thanks for the review. @vfedonkin can you create another PR with the changes requested.
